torch/ao/quantization/dynamic_tracing/auto_trace.py

```diff
@@ -0,0 +1,554 @@
+        2. executes the original module forward
+        3. calls `_auto_quant_state.outputs_prepare_hook`
+
+        Otherwise, calls the original module forward.
+        """
+
+        __interception_module__ = True
+
+        def __call__(self, *args, **kwargs):
+            new_args = map_aggregate(args, convert_to_interception_proxy)
+            new_kwargs = map_aggregate(kwargs, convert_to_interception_proxy)
+            orig_module_call = torch.nn.Module.__call__
+            orig_nn_sequential_forward = torch.nn.Sequential.forward
+
+            def record_module(self, *args, **kwargs):
+
+                if enable_logging:
+                    logger.debug(f"record_module: {type(self)}")
+
+                nonlocal cur_module
+                old_module = cur_module
+                cur_module = self
+                try:
+                    parent_module = module_stack[-1] if len(module_stack) else None
+                    module_stack.append(self)
+                    fqn = module_id_to_fqn.get(id(self), None)
+
+                    if enable_logging:
+                        fqn = module_id_to_fqn.get(id(self), None)
+                        logger.debug(f"\nstarting fqn {fqn}")
+
+                    hook_type = get_module_hook_type(parent_module, cur_module)
+
+                    if hook_type is HookType.OP_HOOKS:
+                        assert parent_module is not None
+                        parent_qstate = parent_module._auto_quant_state
+                        assert isinstance(parent_qstate, AutoQuantizationState)
+                        # before hooks
+                        if not first_call:
+                            parent_qstate.validate_cur_op(cur_module)
+                        args, kwargs = parent_qstate.op_prepare_before_hook(
+                            cur_module, args, kwargs, first_call, qtensor_id,
+                            fqn, cur_module)
+
+                        # original forward
+                        output = orig_module_call(self, *args, **kwargs)
+
+                        # after hooks
+                        # TODO is it correct to call_cur_module twice here?
+                        output = parent_qstate.op_prepare_after_hook(
+                            cur_module, output, args, first_call, qtensor_id,
+                            cur_module)
+                        parent_qstate.mark_cur_op_complete(cur_module)
+
+                    elif hook_type is HookType.MODULE_IO_HOOKS:
+                        # TODO(future PR): add inputs io hook
+
+                        cur_qstate = cur_module._auto_quant_state
+                        cur_qstate.validate_is_at_first_idx()
+
+                        # original forward
+                        output = orig_module_call(self, *args, **kwargs)
+
+                        # after hooks
+                        assert isinstance(cur_qstate, AutoQuantizationState)
+                        output = cur_qstate.outputs_prepare_hook(
```

Comment:
optional nit: can we change this to a more descriptive name? looks like this is not inserting observers, but sets the qtensor_info for the output to be torch.float

Reply:
I understand the feedback, let me schedule some time to discuss in person.  The current naming is generic on purpose, to keep the quantization logic contained in `AutoQuantizationState` and have the caller only call generic hooks.  We can discuss if that's the best approach.